flow/activities/maintenance_activity.go

```diff
@@ -0,0 +1,292 @@
+)
+
+const (
+	mirrorStateBackup   = "backup"
+	mirrorStateRestored = "restore"
+)
+
+type MaintenanceActivity struct {
+	CatalogPool    *pgxpool.Pool
+	Alerter        *alerting.Alerter
+	TemporalClient client.Client
+}
+
+func (a *MaintenanceActivity) GetAllMirrors(ctx context.Context) (*protos.MaintenanceMirrors, error) {
+	rows, err := a.CatalogPool.Query(ctx, `
+	select distinct on(name)
+	  id, name, workflow_id,
+	  created_at, coalesce(query_string, '')='' is_cdc
+	from flows
+	`)
+	if err != nil {
+		return &protos.MaintenanceMirrors{}, err
+	}
+
+	maintenanceMirrorItems, err := pgx.CollectRows(rows, func(row pgx.CollectableRow) (*protos.MaintenanceMirror, error) {
+		var info protos.MaintenanceMirror
+		var createdAt time.Time
+		err := row.Scan(&info.MirrorId, &info.MirrorName, &info.WorkflowId, &createdAt, &info.IsCdc)
+		info.MirrorCreatedAt = timestamppb.New(createdAt)
+		return &info, err
+	})
+	return &protos.MaintenanceMirrors{
+		Mirrors: maintenanceMirrorItems,
+	}, err
+}
+
+func (a *MaintenanceActivity) getMirrorStatus(ctx context.Context, mirror *protos.MaintenanceMirror) (protos.FlowStatus, error) {
+	return shared.GetWorkflowStatus(ctx, a.TemporalClient, mirror.WorkflowId)
+}
+
+func (a *MaintenanceActivity) WaitForRunningSnapshots(ctx context.Context) (*protos.MaintenanceMirrors, error) {
+	mirrors, err := a.GetAllMirrors(ctx)
+	if err != nil {
+		return &protos.MaintenanceMirrors{}, err
+	}
+
+	slog.Info("Found mirrors for snapshot check", "mirrors", mirrors, "len", len(mirrors.Mirrors))
+
+	for _, mirror := range mirrors.Mirrors {
+		lastStatus, err := a.checkAndWaitIfSnapshot(ctx, mirror, 2*time.Minute)
+		if err != nil {
+			return &protos.MaintenanceMirrors{}, err
+		}
+		slog.Info("Finished checking and waiting for snapshot",
+			"mirror", mirror.MirrorName, "workflowId", mirror.WorkflowId, "lastStatus", lastStatus.String())
+	}
+	slog.Info("Finished checking and waiting for all mirrors to finish snapshot")
+	return mirrors, nil
+}
+
+func (a *MaintenanceActivity) checkAndWaitIfSnapshot(
+	ctx context.Context,
+	mirror *protos.MaintenanceMirror,
+	logEvery time.Duration,
+) (protos.FlowStatus, error) {
+	// In case a mirror was just kicked off, it shows up in the running state, we wait for a bit before checking for snapshot
+	if mirror.MirrorCreatedAt.AsTime().After(time.Now().Add(-30 * time.Second)) {
+		slog.Info("Mirror was created less than 30 seconds ago, waiting for it to be ready before checking for snapshot",
+			"mirror", mirror.MirrorName, "workflowId", mirror.WorkflowId)
+		time.Sleep(30 * time.Second)
+	}
+	mirrorStatus, err := a.getMirrorStatus(ctx, mirror)
+	if err != nil {
+		return mirrorStatus, err
+	}
+
+	slog.Info("Checking and waiting if mirror is snapshot", "mirror", mirror.MirrorName, "workflowId", mirror.WorkflowId, "status",
+		mirrorStatus.String())
+	if mirrorStatus != protos.FlowStatus_STATUS_SNAPSHOT && mirrorStatus != protos.FlowStatus_STATUS_SETUP {
+		return mirrorStatus, nil
+	}
```

Comment:
```suggestion
```

Reply:
was planning to check before the loop and log, but I think it is fine anyways